### ISSUE TITLE
Argument-mode parsing: Why is a subexpression at the start of an unquoted token treated differently than a variable reference?

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#3217]
This is a primarily a design rationale question:

The issue is that the rules are hard to remember, and perhaps knowing _why_ they are the way they are helps with that.

If these work:

    $a = 4
    Write-Output $a/H         # -> '4/H'
    Write-Output H/$a         # -> 'H/4'
    Write-Output H/$(2 + 2)   # -> 'H/4'

why shouldn't `Write-Output $(2 + 2)/H` expand to `'4/H'` too, instead of being treated as _2_ arguments, `4` and `'/H'`?

Why is a subexpression at the start treated differently than a variable reference?

Also, why is `Write-Output .$HOME` treated as _2_ arguments (only happens if the `.` is directly followed by a `$`)?

For a full discussion, please see [this Stack Overflow question](http://stackoverflow.com/q/42100211/45375).

**Comment by lzybkr** [comment]
@BrucePay or @JamesWTruher would be the best folks to answer this.

**Comment by JamesWTruher** [comment]
if we look at what the parameter binder believes it's arguments are we can understand the output better:
```
PS> trace-command *param* { write-output $(2+2)/H } -pshost                                                                                                                                                             
DEBUG: ParameterBinderController Information: 0 :  WriteLine   Argument count: 2
<snip>
4
/H
```
vs
```
PS> trace-command *param* { write-output H/$(2+2) } -pshost                                                                                                                                                             
DEBUG: ParameterBinderController Information: 0 :  WriteLine   Argument count: 1
<snip>
H/4
```
So, by the time the arguments get to the parameter binder it's already either broken up (or not) the tokens. There's definitely a difference between a variable reference `$a` and a subexpression `$(2+2)` so i would expect a difference in the behavior between `write-output $(2+2)/H` and `write-output $a/H`.

I _suspect_ that it's the location of the obvious `string` element which causes the behavioral change which changes it from two arguments to one. I haven't tracked that down though. 
`write-output H/...` is an obvious string (since that's how it starts), pretty much anything that follows it which isn't a token separator (`space` for example) is just going to be tacked on the end.

As for `write-output .$HOME` - remember that the `.` serves two functions, one as a invocation operator and another as a property reference operator, depending on context. In this case, we've definitely got 2 tokens `.` (the invocation operator) and `$HOME`, so we emit two lines. 
FWIW, `Write-Host` sees these two tokens as well:
```
Write-Host .$HOME
. /Users/james
```
note the space. Why the difference? The difference between `write-host` and `write-object` is the parameter type for the input object. in `write-host` the type is `object` and in `write-output` the parameter type is `PSObject[]`.

**Comment by mklement0** [comment]
Thanks for that, @JamesWTruher.

To provide more motivation for my question, which is primarily about the _user experience_, not the technical underpinnings:

* It is obscure to have behavior depend on the order of tokens in this case, especially since variable references and subexpressions are _not_ treated differently when explicitly embedded inside `"..."`

  * `Write-Output "$(2 + 2)/H"` yields only 1 argument.
  * In POSIX-like shells, a command/arithmetic substitution (the equivalent of a subexpression in PS) _does_ work as the first token:  
`echo $(( 2 + 2 ))/H` yields 1 argument, `4/H`
  * That said, (the expanded results of) unquoted tokens in POSIX-like shells are subject to shell expansions that are often unwanted, leading to the annoyance that double quotes _must_ be used everywhere (and beginners forgetting them is a constant source of bugs).
  * PowerShell generally frees us from that annoyance, so it's unfortunate that it doesn't do so in this case (because it results in multiple arguments).  
 
  * Generally, _remembering_ when it works and when it doesn't, in the absence of an intuitive reason, is a challenge and, I suspect, will be an ongoing source of confusion.

* The same obscure  behavior applies to juxtaposing differently quoted substrings (which are commonly used in POSIX-like shells):

   * `Write-Output bare"I'm quoted"` -> _1_ argument, because the 1st token is _unquoted_:  
`bareI'm quoted`
   * `Write-Output "I'm quoted"bare` -> _2_ arguments, because the 1st token is _quoted_:  
`I'm quoted` and `bare`

* As for the `.$HOME` case: Given that the token is a _non-initial_ token parsed in _argument mode_, it _shouldn't_ be interpreted as the invocation operator.  
If unquoted tokens are to be treated like _expandable strings_ (i.e., as if the token were explicitly enclosed in `"..."`, "_unless it begins with one of the following special characters: dollar sign (`$`), at sign (`@`), single quotation mark (`'`), double quotation mark (`"`), or an opening parenthesis (`(`)_", according to `Get-Help about_Parsing`), then `.$HOME` should work the same way as, say, `x$HOME`.

I suspect the answer is that changing the behaviors described is way too late and risky, but perhaps the pitfalls of the existing behavior can be _documented_ somewhere.

**Comment by iRon7** [comment]
I came across this issue as I created a similar [Unquoted Strings](https://github.com/PoshCode/PowerShellPracticeAndStyle/issues/146) request at the [The PowerShell Best Practices and Style Guide](https://github.com/PoshCode/PowerShellPracticeAndStyle) a few weeks ago without any activity. I have also created a similar statckoverflow question: [Unquoted key rules / best practice](https://stackoverflow.com/q/62754771/1701026), as I got lost in the (open and close) cascaded issues an dwhether this is already covered or even answered. And if not, to which issue I should add my findings (or create a new one?).
